firefly/kernel/intel64/cpu/apic/apic.cpp

```diff
@@ -0,0 +1,123 @@
+#include "firefly/intel64/cpu/apic/apic.hpp"
+
+#include <cstring>
+
+#include "firefly/drivers/ports.hpp"
+#include "firefly/intel64/acpi/acpi.hpp"
+#include "firefly/intel64/cpu/cpu.hpp"
+#include "firefly/limine.hpp"
+#include "firefly/memory-manager/primary/buddy.hpp"
+#include "firefly/memory-manager/primary/primary_phys.hpp"
+#include "firefly/timer/timer.hpp"
+#include "libk++/bits.h"
+
+namespace firefly::kernel::apic {
+using core::acpi::Acpi;
+
+frg::manual_box<Apic> apicSingleton;
+frg::manual_box<ApicTimer> apicTimerSingleton;
+
+Apic& Apic::accessor() {
+    return *apicSingleton;
+}
+
+ApicTimer& ApicTimer::accessor() {
+    return *apicTimerSingleton;
+}
+
+// Write to the APIC
+void Apic::write(uint32_t offset, uint32_t value) {
```

Comment:
I'm thinking that most of this would be better off just in a class, and whoever needs a timer would inherit this class y'know?

Like...
```c++
enum TimerPreference : int {
   Apic, Hpet, Pit, Any /* Any => Apic, Hpet, Pit in that order. */
};
class GenericTimer<typename TimerPreference> : ApicTimer, HpetTimer, PitTimer /* expose some protected members like registers idk. We can change this stuff later on */ {
     void initialize() {
          // try to initialize apic, hpet or pit unless a timer preference is set. 
          // if a timer preference is NOT 'Any', try to initialize that first
     }

     void sleep(seconds) {
         timer_selection.sleep(seconds);
     }

     TimerPreference preference; // preferred timer (hept, pit, apic)
     frg::tuple<ApicTimer, HpetTimer, PitTimer> timer_selection; // Which timer to use. This might seem a little weird or whatever, I'm just in a rush atm
     uint8_t timer_selection_index {0}; // this would be an index into timer_selection
};
```

Reply:
timer_selection would be of type TimerPreference

Reply:
I'd say not really, because I wouldn't want a sleep() for every timer, we should use Oneshot lapic timer just set to 1ms or smth so we can just count ms and have a list of things wanting to sleep, probably a better solution than a sleep for everything.

Reply:
Or, PIT set to 1ms for example

Reply:
Then if we want task switches every 20 ms, just check if 20 ticks passed since then

Reply:
Well yeah, then have a setTimerIrq(duration, handler) 

What if the machine doesn't have hpet but you insist on using hpet?